## MVP Know Your Transaction (KYT) - Transaction Risk Classification Engine

### Project Overview

This notebook presents a comprehensive implementation of a Transaction Risk Classification Engine for Anti-Money Laundering (AML) compliance in cryptocurrency transactions. The project addresses the critical need for risk assessment of Bitcoin transactions by combining traditional AML indicators with blockchain-specific risk factors.

### Domain Context: Financial AML for Transactions

#### Core Domain Definition
Anti-Money Laundering (AML) for transactions encompasses the comprehensive framework of laws, regulations, procedures, and technological solutions designed to prevent criminals from disguising illegally obtained funds as legitimate income through the global financial system. This domain includes detection, prevention, and reporting of money laundering, terrorist financing, tax evasion, market manipulation, and misuse of public funds.

### Problem Definition: Transaction Risk Classification Engine

#### Problem Statement
Develop a system that assigns risk classifications to cryptocurrency transactions in real-time, integrating traditional AML indicators with blockchain-specific risk factors including wallet clustering, transaction graph analysis, and counterparty reputation scoring.

#### Technical Requirements
- **Problem Type**: Classification 
- **Processing Speed**: Sub-second analysis for high-frequency transactions
- **Difficulty Level**: High - requires complex multi-dimensional data processing
- **Output Format**: Risk binary classification (illicit = 1 /licit = 2)

#### Data Landscape
The system can processes multiple data dimensions:
- Transaction metadata (amounts, timestamps, fees)
- Wallet addresses and clustering information
- Transaction graph relationships and network topology
- Counterparty databases and reputation scores
- Sanctions lists and regulatory databases
- Temporal patterns and behavioral baselines

### References

This notebook implementation is based on the comprehensive research and analysis conducted during the project development phase. The following reference documents were used in the composition of this initial description:

- **Domain Research**: [current-domain.md](domains/current-domain.md) - Contains detailed market analysis, regulatory framework research, and commercial viability assessment for the Financial AML domain;
- **Problem Analysis**: [current-problem.md](problems/current-problem.md) - Provides comprehensive problem refinement, technical requirements analysis, and solution approach evaluation;
- **Dataset Evaluation**: [current-dataset.md](datasets/current-dataset.md) - Documents dataset selection criteria, suitability scoring, and detailed feature analysis for the Elliptic dataset;
- **Dataset Analysis & Preprocessing**: [dataset-analysis-and-preprocessing.ipynb](datasets/scripts/dataset-analysis-and-preprocessing.ipynb) - Comprehensive Jupyter notebook containing Elliptic dataset download, exploratory data analysis, and ML preparation steps;

These reference documents contain the foundational research that informed the technical approach, feature engineering strategy, and implementation decisions reflected in this notebook.

---

This notebook serves as the primary entry point for the MVP KYT implementation and it can run independently, providing both technical implementation and business context for real cryptocurrency transaction risk assessment.

### Import Libraries

Comprehensive installation and import of all required libraries for machine learning procedures.

In [87]:
!pip install azure-storage-blob


from pathlib import Path
from azure.storage.blob import BlobServiceClient


"""
Utility functions for dataset processing and Azure blob storage operations.
"""
class AzureBlobDownloader:
    """
    Azure Blob Storage downloader class for managing dataset downloads.

    This class encapsulates Azure blob operations and maintains connection state
    for efficient dataset management operations.
    """

    def __init__(self, account_url, container_name):
        """
        Initialize Azure blob downloader.

        Args:
            account_url (str): Azure storage account URL
            container_name (str): Name of the blob container

        Raises:
            Exception: If connection to Azure fails or Azure SDK not available
        """
        try:
            self.account_url = account_url
            self.container_name = container_name
            self.blob_service_client = BlobServiceClient(account_url=account_url)
            self.container_client = self.blob_service_client.get_container_client(container_name)

        except Exception as e:
            raise Exception(f"Failed to initialize Azure Blob connection: {e}")


    def download_documents(self, project_folder, document_folder, base_path="../"):
        """
        Download dataset from Azure Blob Storage.

        Args:
            dataset_name: Name of the dataset directory in blob storage
            base_path: Local base path for downloads (default: "../")

        Returns:
            bool: True if download successful, False otherwise
        """
        try:
            original_dir = Path(base_path) / project_folder 
            specific_dir = original_dir / document_folder

            original_dir.mkdir(exist_ok=True)
            specific_dir.mkdir(exist_ok=True)

            remote_path = f"{project_folder}/{document_folder}/"
            downloaded_files = 0

            for blob in self.container_client.list_blobs(name_starts_with=remote_path):
                blob_client = self.container_client.get_blob_client(blob.name)
                local_file_path = Path(base_path) / blob.name
                local_file_path.parent.mkdir(parents=True, exist_ok=True)

                blob_data = blob_client.download_blob()
                with open(local_file_path, "wb") as download_file:
                    download_file.write(blob_data.readall())
                downloaded_files += 1

            print(f"Successfully downloaded {downloaded_files} files from Azure Blob Storage")
            return True

        except Exception as e:
            print(f"Failed to download from Azure Blob Storage: {e}")
            return False

In [88]:
import os
import warnings
import joblib
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import make_scorer 
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score 
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import StackingClassifier
from sklearn.metrics import f1_score
import optuna
from optuna.samplers import TPESampler

# Suppress all warnings
warnings.filterwarnings('ignore')
optuna.logging.set_verbosity(optuna.logging.WARNING)

azureClient = AzureBlobDownloader("https://stmvppos.blob.core.windows.net", "mvpkytsup")

### Loading pre-processed datasets

Let's load the pre-processed and compressed data from remote and local sources. The dataset is a composition of cryptocurrency transactions in the Bitcoin blockchain. 

It has 166 features, of which 92 features represent local transaction information and another 72 features represent aggregated information from one-hop neighboring transactions (directly linked transactions). Thus, the dataset already has curated information about the relationships between transactions. 

This is very important because money laundering and fraud patterns often involve coordinated transaction clusters and neighborhood patterns.

In [89]:
# Define paths
dataset_dir = Path("elliptic_bitcoin_dataset")
root_dir = Path("./datasets")
processed_dir = root_dir / "processed" / dataset_dir
root_dir.mkdir(exist_ok=True)

# Download processed data from Azure if not present locally
if not processed_dir.exists() or not any(processed_dir.iterdir()):
    print(f"Local processed directory is empty. Downloading from Azure...")
    azureClient.download_documents("datasets/processed", dataset_dir.name, base_path="./")

# The complete dataset already pre-processed 
df_complete = pd.read_hdf(processed_dir / "df_complete.h5", key="df_complete")
print(f"Loaded from HDF5: {df_complete.shape} - All subsequent operations will use compressed data")

# The filtered labeled dataset already pre-processed
df_labeled = pd.read_hdf(processed_dir / "df_labeled.h5", key="df_labeled")
print(f"Loaded from HDF5: {df_labeled.shape} - All subsequent operations will use compressed data")

# The filtered unlabeled dataset already pre-processed
df_unlabeled = pd.read_hdf(processed_dir / "df_unlabeled.h5", key="df_unlabeled")
print(f"Loaded from HDF5: {df_unlabeled.shape} - All subsequent operations will use compressed data")

# The edges dataset that maps relationships between transaction nodes
df_edges = pd.read_hdf(processed_dir / "df_edges.h5", key="df_edges")
print(f"Loaded from HDF5: {df_edges.shape} - All subsequent operations will use compressed data")

# Summary of all datasets
print(f"\n📊 Dataset Summary:")
print(f"  - Features: {df_complete.shape[0]:,} transactions × {df_complete.shape[1] -2} features")
print(f"  - Labeled: {df_labeled.shape[0]:,} transactions")
print(f"  - Unlabeled: {df_unlabeled.shape[0]:,} transactions")
print(f"  - Edges: {df_edges.shape[0]:,} transaction relationships")

Loaded from HDF5: (203769, 168) - All subsequent operations will use compressed data
Loaded from HDF5: (46564, 168) - All subsequent operations will use compressed data
Loaded from HDF5: (157205, 168) - All subsequent operations will use compressed data
Loaded from HDF5: (234355, 2) - All subsequent operations will use compressed data

📊 Dataset Summary:
  - Features: 203,769 transactions × 166 features
  - Labeled: 46,564 transactions
  - Unlabeled: 157,205 transactions
  - Edges: 234,355 transaction relationships


### Machine Learning Strategy

Let's apply machine learning techniques to the labeled dataset portion using supervised learning, and apply the prediction to the unknown unlabeled dataset portion in order to establish a performance baseline for future improvements.

Following these steps:

1. Define overall parameters and make data splits;
2. Define all training models to be used;
3. Define all pipelines to be used during training;
4. Define model parameter distributions for a grid search approach;
5. Define the score function to be used during training;
6. Execute the training;
7. Save all resulting models;
8. Validate all models and select the best models;
9. Use best models to predict unknown data;

**1. Define overall parameters and make data splits**

After splitting the labeled dataset into training and validation (test) sets, let's prepare the training dataset for training and validation using the stratified approach, which generates a fixed number of splits (folds) for the dataset following a fixed proportion of train/test samples. 

The main idea of this approach is to guarantee training without bias toward a specific dataset split because it maintains the same class proportion for each split (fold) generated. We use this technique in labeled and supervised learning, assuming that the dataset's pattern does not have significant changes over time.

In this specific dataset, we consider the timestamp as a grouping factor for transactions but not as a changing factor for the dataset's pattern over time.

In [90]:
# Defining overall parameters
random_seed = 4354 # PARAMETER: random seed
test_size_split = 0.20 # PARAMETER: test set size
n_stratified_splits = 2 # PARAMETER: number of folds
n_pca_components = 0.95 # PARAMETER: PCA components to keep

np.random.seed(random_seed)

# Prepare data 
x_labeled = df_labeled.drop(['class', 'txId'], axis=1)
y_labeled = df_labeled['class']  # Binary target

# Split the data into training and testing set
X_train, X_test, y_train, y_test = train_test_split(x_labeled, y_labeled,
    test_size=test_size_split, 
    shuffle=True, 
    random_state=random_seed, 
    stratify=y_labeled) # stratified holdout

# Cross-validation setup to be applied in the training set 
cv = StratifiedKFold(n_splits=n_stratified_splits, 
                     shuffle=True, 
                     random_state=random_seed)

**2. Define all training models to be used**

Let's define which models to use during training, selecting diverse ML algorithms covering linear, tree-based, probabilistic, and ensemble methods. 

Different algorithms capture different transaction patterns; ensemble methods reduce overfitting risk and can improve individual model performance.

In [91]:
# Defining the individual models
reg = ('LR', LogisticRegression())
knn = ('KNN', KNeighborsClassifier())
cart = ('CART', DecisionTreeClassifier())
nav = ('NB', GaussianNB())
svm = ('SVM', SVC())

# Defining ensemble models
bagging = ('Bag', BaggingClassifier())
forest = ('RF', RandomForestClassifier())
extra = ('ET', ExtraTreesClassifier())
ada = ('Ada', AdaBoostClassifier())
gradient = ('GB', GradientBoostingClassifier())

**3. Define all pipelines to be used during training**

Let's define which ML pipelines to use during training, by configuring which pre-processing operations and models will be used in the training process. Pipelines also help avoid data leakage by ensuring that feature transformation is applied only to the training dataset portion.          

For feature transformation, the standard scaler was used to ensure the best scale for feature values. By normalizing all values to a common metric, it reduces bias toward feature magnitude and enables subsequent operations to capture important patterns between features without being influenced mainly by their magnitude. All features contribute equally to pattern detection. This is especially important in finance because the difference between feature scales can be significant.

For feature dimensionality reduction, Principal Component Analysis (PCA) was used to reduce from 166 to only 59 features. We aim to reduce feature dimensionality while maximizing the dissimilarity of the original dataset, thus extracting the most discriminative features and improving model training performance. It must be applied after the standard scaler to avoid the mentioned magnitude bias, and should be used in datasets that have a large quantity of features.

In [92]:
# Creating the pipelines
pipelines = []
std  = ('std', StandardScaler())  # Standardization

# Defining the pipelines, for future experimentation 
#pipelines.append(('LR', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), reg])))  
#pipelines.append(('NB', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), nav])))  
#pipelines.append(('KNN', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), knn])))
#pipelines.append(('CART', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), cart])))
#pipelines.append(('SVM', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), svm])))
#pipelines.append(('Bag', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), bagging])))
#pipelines.append(('RF', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), forest])))
#pipelines.append(('ET', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), extra])))
pipelines.append(('Ada', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), ada])))
pipelines.append(('GB', Pipeline([std, ('pca', PCA(n_components=n_pca_components)), gradient])))

# ENSEMBLE: SVM + KNN Voting Soft
voting_soft = ('Vote-Soft', Pipeline([
    std, 
    ('pca', PCA(n_components=n_pca_components)),
    ('voting', VotingClassifier(
        estimators=[
            ('svm', SVC(probability=True)),  # Enable probability for soft voting
            ('knn', KNeighborsClassifier())
        ],
        voting='soft',
        weights=[0.6, 0.4]  # Weight SVM higher (better performer)
    ))
]))

# ENSEMBLE: SVM + KNN Voting Hard
voting_hard = ('Vote-Hard', Pipeline([
    std,
    ('pca', PCA(n_components=n_pca_components)),
    ('voting', VotingClassifier(
        estimators=[
            ('svm', SVC()),
            ('knn', KNeighborsClassifier())
        ],
        voting='hard'
    ))
]))

# ENSEMBLE: SVM + KNN Stacking
stacking = ('Stack', Pipeline([
    std,
    ('pca', PCA(n_components=n_pca_components)),
    ('stacking', StackingClassifier(
        estimators=[
            ('svm', SVC(probability=True)),
            ('knn', KNeighborsClassifier())
        ],
        final_estimator=LogisticRegression(),  # Meta-learner
        cv=5,
        passthrough=False  # Only use base predictions, not original features
    ))
]))

pipelines.append(voting_soft)
pipelines.append(voting_hard)
pipelines.append(stacking)

# ENSEMBLE: Bagging with SVM base estimator
bag_svm = ('Bag-SVM', Pipeline([
    std,
    ('pca', PCA(n_components=n_pca_components)),
    ('bagging', BaggingClassifier(
        estimator=SVC(probability=True),
        n_estimators=10,
        max_samples=0.8,
        max_features=0.8,
        bootstrap=True,
        random_state=random_seed
    ))
]))

# ENSEMBLE: Bagging with KNN base estimator
bag_knn = ('Bag-KNN', Pipeline([
    std,
    ('pca', PCA(n_components=n_pca_components)),
    ('bagging', BaggingClassifier(
        estimator=KNeighborsClassifier(),
        n_estimators=10,
        max_samples=0.8,
        max_features=0.8,
        bootstrap=True,
        random_state=random_seed
    ))
]))

# ENSEMBLE: AdaBoost with SVM base estimator
ada_svm = ('Ada-SVM', Pipeline([
    std,
    ('pca', PCA(n_components=n_pca_components)),
    ('adaboost', AdaBoostClassifier(
        estimator=SVC(probability=True, kernel='linear'),  # Linear kernel required for AdaBoost
        n_estimators=50,
        learning_rate=1.0,
        random_state=random_seed
    ))
]))

# ENSEMBLE: AdaBoost with KNN base estimator
ada_knn = ('Ada-KNN', Pipeline([
    std,
    ('pca', PCA(n_components=n_pca_components)),
    ('adaboost', AdaBoostClassifier(
        estimator=KNeighborsClassifier(),
        n_estimators=50,
        learning_rate=1.0,
        algorithm='SAMME',  # SAMME works better with KNN
        random_state=random_seed
    ))
]))

pipelines.append(bag_svm)
pipelines.append(bag_knn)
pipelines.append(ada_svm)
pipelines.append(ada_knn)

**4. Define model parameter distributions for a grid search approach**

Let's prepare the parameter distributions for a random grid search, by using a distribution of possible parameter values, so that the training phase can explore the best performance models also from the perspective of their hyperparameters. This is a better solution than the common grid search approach because it can explore a broader hyperparameter space and is often faster. 

Three types of parameter distributions were used:

- uniform: Continuous values with equal probability across a range, used when all values in the range are equally valid;
- loguniform: Continuous values on logarithmic scale (exponential distribution), used when smaller values are often better;
- randint: Discrete integer values with equal probability, used for discrete hyperparameters.

In [93]:
# Define parameter distributions using Optuna trial lambdas
# Enhanced with comprehensive documentation
# Optimized for financial/cryptocurrency transaction risk classification

param_distributions = {
    'LR': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        # n_components: Variance ratio to retain (0.90-0.99 recommended for financial data)
        # Lower values = more dimensionality reduction = faster training = potential information loss
        # Higher values = retain more variance = preserve patterns = longer training
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),

        # whiten: Whitening transformation for decorrelation
        # True: Transform components to have unit variance (helps some models)
        # False: Keep original scale after PCA transformation
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),

        # svd_solver: Algorithm for singular value decomposition
        # 'auto': Automatic selection based on data shape and n_components
        # 'full': Standard SVD, works with variance ratio n_components
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # REGULARIZATION STRENGTH (C): Inverse regularization strength
        # Lower values = stronger regularization = simpler model (prevent overfitting)
        # Higher values = weaker regularization = more complex model
        # Log-uniform for exponential search across orders of magnitude
        # Critical for financial data: balance between model complexity and generalization
        'LR__C': lambda trial: trial.suggest_float('LR__C', 1e-4, 1e2, log=True),

        # OPTIMIZATION SOLVER: Algorithm for weight optimization
        # 'lbfgs': Fast for small datasets, good convergence, handles L2/none penalty
        # 'newton-cg': Robust for large datasets, handles L2/none penalty
        # 'sag': Stochastic Average Gradient, fast for large datasets
        # 'saga': Supports all penalties, good for sparse features (financial data)
        # Removed 'liblinear': slower for datasets > 10K samples
        'LR__solver': lambda trial: trial.suggest_categorical('LR__solver', ['lbfgs', 'newton-cg', 'sag', 'saga']),

        # REGULARIZATION TYPE: Controls feature selection and overfitting
        # 'l2': Ridge regression, keeps all features, reduces coefficients
        # None: No regularization, may overfit with many features
        # Removed 'l1': Lasso causes issues with solver compatibility
        # Financial context: L2 better for correlated transaction features
        'LR__penalty': lambda trial: trial.suggest_categorical('LR__penalty', ['l2', None]),

        # MAXIMUM ITERATIONS: Convergence limit for optimization
        # Higher values ensure convergence but increase training time
        # Financial data often needs more iterations due to class imbalance
        # 2000+ recommended for 46K+ samples to avoid convergence warnings
        'LR__max_iter': lambda trial: trial.suggest_categorical('LR__max_iter', [1000, 2000, 5000]),

        # CONVERGENCE TOLERANCE: Stopping criteria precision
        # Lower values = more precise convergence = longer training time
        # Financial models need precise convergence for regulatory compliance
        # 1e-6: High precision, 1e-4: Fast convergence
        'LR__tol': lambda trial: trial.suggest_float('LR__tol', 1e-6, 1e-3, log=True)
    },

    'KNN': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # NUMBER OF NEIGHBORS: Core hyperparameter for KNN algorithm
        # Lower values = more complex decision boundary = higher variance
        # Higher values = smoother decision boundary = higher bias
        # Financial context: 5-15 often optimal for transaction classification
        # Odd numbers prevent ties in binary classification
        'KNN__n_neighbors': lambda trial: trial.suggest_int('KNN__n_neighbors', 3, 20),

        # WEIGHT FUNCTION: How neighbors influence prediction
        # 'uniform': All neighbors weighted equally
        # 'distance': Closer neighbors have higher influence
        # Financial context: 'distance' often better for transaction patterns
        'KNN__weights': lambda trial: trial.suggest_categorical('KNN__weights', ['uniform', 'distance']),

        # DISTANCE METRIC: How to measure similarity between transactions
        # 'euclidean': Standard L2 distance, good for continuous features
        # 'manhattan': L1 distance, robust to outliers (good for financial data)
        # 'minkowski': Generalization, controlled by 'p' parameter
        'KNN__metric': lambda trial: trial.suggest_categorical('KNN__metric', ['euclidean', 'manhattan', 'minkowski']),

        # MINKOWSKI POWER: Only used when metric='minkowski'
        # p=1: Manhattan distance, p=2: Euclidean distance
        # Financial data: p=1 often better due to outlier robustness
        'KNN__p': lambda trial: trial.suggest_int('KNN__p', 1, 2)
    },

    'CART': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # MAXIMUM TREE DEPTH: Primary overfitting control
        # Lower values = simpler tree = less overfitting = higher bias
        # Financial context: 5-15 often optimal for interpretability vs performance
        # Too deep trees memorize transactions instead of learning patterns
        'CART__max_depth': lambda trial: trial.suggest_int('CART__max_depth', 3, 19),

        # MINIMUM SAMPLES TO SPLIT: Prevents splitting on small sample sizes
        # Higher values = more conservative splits = less overfitting
        # Financial context: 10-50 good for 46K+ dataset to ensure robust splits
        'CART__min_samples_split': lambda trial: trial.suggest_int('CART__min_samples_split', 10, 49),

        # MINIMUM SAMPLES PER LEAF: Ensures leaf nodes have sufficient samples
        # Higher values = smoother predictions = less overfitting
        # Critical for financial data: prevents decisions based on few transactions
        'CART__min_samples_leaf': lambda trial: trial.suggest_int('CART__min_samples_leaf', 5, 19),

        # SPLIT QUALITY MEASURE: Criterion for evaluating split quality
        # 'gini': Gini impurity, faster computation
        # 'entropy': Information gain, potentially better separation
        # Financial context: both work well, entropy slightly better for imbalanced data
        'CART__criterion': lambda trial: trial.suggest_categorical('CART__criterion', ['gini', 'entropy'])
    },

    'NB': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # VARIANCE SMOOTHING: Laplace smoothing for numerical stability
        # Prevents zero probabilities when features have zero variance
        # Log-uniform for exponential search across precision levels
        # Financial context: 1e-9 to 1e-6 range good for PCA-transformed features
        'NB__var_smoothing': lambda trial: trial.suggest_float('NB__var_smoothing', 1e-12, 1e-6, log=True),

        # CLASS PRIORS: Prior probabilities for each class
        # None: Learn priors from training data (recommended)
        # Custom priors could be set based on known illicit transaction rates
        'NB__priors': lambda trial: trial.suggest_categorical('NB__priors', [None])
    },

    'SVM': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # REGULARIZATION PARAMETER: Trade-off between margin and misclassification
        # Lower C = wider margin = more regularization = simpler model
        # Higher C = narrower margin = less regularization = complex model
        # Log-uniform critical for SVM: performance varies across orders of magnitude
        # Financial context: Often needs tuning from 0.01 to 1000
        'SVM__C': lambda trial: trial.suggest_float('SVM__C', 1e-2, 1e3, log=True),

        # KERNEL FUNCTION: Maps features to higher-dimensional space
        # 'rbf': Radial basis function, good for non-linear financial patterns
        # 'poly': Polynomial kernel, can capture feature interactions
        # 'sigmoid': Tanh kernel, neural network-like behavior
        # Removed 'linear': redundant with LogisticRegression for PCA features
        'SVM__kernel': lambda trial: trial.suggest_categorical('SVM__kernel', ['rbf', 'poly', 'sigmoid']),

        # KERNEL COEFFICIENT: Controls kernel shape and influence
        # 'scale': 1/(n_features * X.var()) - good default for normalized data
        # 'auto': 1/n_features - simpler scaling
        # Log-uniform range would be loguniform(1e-6, 1e-1) for custom values
        'SVM__gamma': lambda trial: trial.suggest_categorical('SVM__gamma', ['scale', 'auto'])
    },

    'RF': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # NUMBER OF TREES: Primary performance vs speed trade-off
        # More trees = better performance = longer training/prediction time
        # Financial context: 100-500 often optimal, diminishing returns after 300
        # Real-time KYT systems need to balance accuracy with inference speed
        'RF__n_estimators': lambda trial: trial.suggest_int('RF__n_estimators', 100, 499),

        # MAXIMUM TREE DEPTH: Individual tree complexity control
        # None = trees grow until pure leaves (may overfit)
        # Specific values prevent overfitting in ensemble
        # Financial context: 10-25 good for transaction complexity
        'RF__max_depth': lambda trial: trial.suggest_int('RF__max_depth', 10, 24),

        # MINIMUM SAMPLES TO SPLIT: Conservative splitting threshold
        # Higher values = more robust trees = better generalization
        # Financial data: 5-20 good for 46K+ samples
        'RF__min_samples_split': lambda trial: trial.suggest_int('RF__min_samples_split', 5, 19),

        # MINIMUM SAMPLES PER LEAF: Leaf node size constraint
        # Prevents overfitting to individual transactions
        # Financial context: 2-10 ensures meaningful leaf nodes
        'RF__min_samples_leaf': lambda trial: trial.suggest_int('RF__min_samples_leaf', 2, 9),

        # FEATURE SUBSET SIZE: Number of features per tree split
        # 'sqrt': sqrt(n_features) ≈ 7 features for 46 total
        # 'log2': log2(n_features) ≈ 6 features for 46 total
        # None: Use all features (may reduce diversity)
        # Financial context: 'sqrt' often optimal for transaction features
        'RF__max_features': lambda trial: trial.suggest_categorical('RF__max_features', ['sqrt', 'log2', None]),

        # BOOTSTRAP SAMPLING: Sample replacement for tree training
        # True: Standard random forest with replacement sampling
        # False: Use entire dataset for each tree (less diversity)
        # Financial context: True recommended for better generalization
        'RF__bootstrap': lambda trial: trial.suggest_categorical('RF__bootstrap', [True, False])
    },

    'ET': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # Extra Trees: Similar to Random Forest but with random splits
        # Generally faster training, sometimes better performance
        # Parameters similar to Random Forest but more aggressive randomization

        'ET__n_estimators': lambda trial: trial.suggest_int('ET__n_estimators', 100, 499),          # Number of trees in ensemble
        'ET__max_depth': lambda trial: trial.suggest_int('ET__max_depth', 10, 24),               # Maximum depth of trees
        'ET__min_samples_split': lambda trial: trial.suggest_int('ET__min_samples_split', 5, 19),        # Min samples to split node
        'ET__min_samples_leaf': lambda trial: trial.suggest_int('ET__min_samples_leaf', 2, 9),         # Min samples at leaf
        'ET__max_features': lambda trial: trial.suggest_categorical('ET__max_features', ['sqrt', 'log2', None]),     # Random feature subset size
        'ET__bootstrap': lambda trial: trial.suggest_categorical('ET__bootstrap', [True, False])                  # Bootstrap sampling toggle
    },

    'GB': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # BOOSTING STAGES: Number of sequential weak learners
        # More stages = better performance = higher overfitting risk
        # Financial context: 100-300 often optimal, early stopping recommended
        'GB__n_estimators': lambda trial: trial.suggest_int('GB__n_estimators', 100, 299),

        # LEARNING RATE: Step size shrinkage for gradient updates
        # Lower rates = more conservative learning = need more estimators
        # Higher rates = aggressive learning = risk of overshooting optimum
        # Log-uniform for exponential search: 0.01-0.3 typical range
        # Financial context: 0.05-0.15 often optimal for stability
        'GB__learning_rate': lambda trial: trial.suggest_float('GB__learning_rate', 1e-2, 3e-1, log=True),

        # INDIVIDUAL TREE DEPTH: Weak learner complexity
        # Gradient boosting uses shallow trees (stumps to 8 levels)
        # Higher depth = stronger individual learners = risk of overfitting
        # Financial context: 3-8 optimal for transaction patterns
        'GB__max_depth': lambda trial: trial.suggest_int('GB__max_depth', 3, 7),

        # SUBSAMPLE FRACTION: Stochastic gradient boosting
        # < 1.0 = use random subset of samples for each tree
        # Reduces overfitting and improves generalization
        # Financial context: 0.7-0.9 good for large transaction datasets
        'GB__subsample': lambda trial: trial.suggest_float('GB__subsample', 0.7, 0.9)
    },

    'Ada': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # ADAPTIVE BOOSTING: Sequential weak learner weighting
        # Focuses on previously misclassified transactions

        # NUMBER OF WEAK LEARNERS: Maximum boosting rounds
        # AdaBoost often needs fewer estimators than Gradient Boosting
        # Financial context: 50-200 sufficient for most transaction patterns
        'Ada__n_estimators': lambda trial: trial.suggest_int('Ada__n_estimators', 50, 199),

        # LEARNING RATE: Weight applied to each weak classifier
        # Lower rates = more conservative = better generalization
        # Higher rates = aggressive = faster convergence but overfitting risk
        # Financial context: 0.5-1.5 range for transaction classification
        'Ada__learning_rate': lambda trial: trial.suggest_float('Ada__learning_rate', 0.5, 1.5),

        # BOOSTING ALGORITHM: AdaBoost variant
        # 'SAMME': Discrete AdaBoost, works with any base classifier
        'Ada__algorithm': lambda trial: trial.suggest_categorical('Ada__algorithm', ['SAMME'])
    },

    'Bag': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # BOOTSTRAP AGGREGATING: Parallel ensemble method
        # Reduces overfitting through averaging multiple models

        'Bag__n_estimators': lambda trial: trial.suggest_int('Bag__n_estimators', 50, 199),          # Number of base estimators

        # SAMPLE FRACTION: Proportion of dataset for each estimator
        # Lower values = more diversity = better generalization
        # Higher values = more stable individual models
        # Financial context: 0.6-0.9 good for transaction data diversity
        'Bag__max_samples': lambda trial: trial.suggest_float('Bag__max_samples', 0.6, 0.9),

        # FEATURE FRACTION: Proportion of features for each estimator
        # Creates feature diversity in ensemble
        # Financial context: 0.7-1.0 to maintain transaction pattern integrity
        'Bag__max_features': lambda trial: trial.suggest_float('Bag__max_features', 0.7, 1.0)
    },
    
    'Vote-Soft': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # SVM hyperparameters for voting ensemble
        'voting__svm__C': lambda trial: trial.suggest_float('voting__svm__C', 1e-2, 1e3, log=True),
        'voting__svm__kernel': lambda trial: trial.suggest_categorical('voting__svm__kernel', ['rbf', 'poly']),
        'voting__svm__gamma': lambda trial: trial.suggest_categorical('voting__svm__gamma', ['scale', 'auto']),
        
        # KNN hyperparameters for voting ensemble
        'voting__knn__n_neighbors': lambda trial: trial.suggest_int('voting__knn__n_neighbors', 3, 20),
        'voting__knn__weights': lambda trial: trial.suggest_categorical('voting__knn__weights', ['uniform', 'distance']),
        'voting__knn__metric': lambda trial: trial.suggest_categorical('voting__knn__metric', ['euclidean', 'manhattan']),
        
        # Voting weights optimization
        'voting__weights': lambda trial: trial.suggest_categorical('voting__weights', [[0.7, 0.3], [0.6, 0.4], [0.5, 0.5], [0.8, 0.2]])
    },
    
    'Vote-Hard': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # SVM hyperparameters for voting ensemble
        'voting__svm__C': lambda trial: trial.suggest_float('voting__svm__C', 1e-2, 1e3, log=True),
        'voting__svm__kernel': lambda trial: trial.suggest_categorical('voting__svm__kernel', ['rbf', 'poly']),
        'voting__svm__gamma': lambda trial: trial.suggest_categorical('voting__svm__gamma', ['scale', 'auto']),
        
        # KNN hyperparameters for voting ensemble
        'voting__knn__n_neighbors': lambda trial: trial.suggest_int('voting__knn__n_neighbors', 3, 20),
        'voting__knn__weights': lambda trial: trial.suggest_categorical('voting__knn__weights', ['uniform', 'distance']),
        'voting__knn__metric': lambda trial: trial.suggest_categorical('voting__knn__metric', ['euclidean', 'manhattan'])
    },
    
    'Stack': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # SVM hyperparameters for stacking
        'stacking__svm__C': lambda trial: trial.suggest_float('stacking__svm__C', 1e-2, 1e3, log=True),
        'stacking__svm__kernel': lambda trial: trial.suggest_categorical('stacking__svm__kernel', ['rbf', 'poly']),
        'stacking__svm__gamma': lambda trial: trial.suggest_categorical('stacking__svm__gamma', ['scale', 'auto']),
        
        # KNN hyperparameters for stacking
        'stacking__knn__n_neighbors': lambda trial: trial.suggest_int('stacking__knn__n_neighbors', 3, 20),
        'stacking__knn__weights': lambda trial: trial.suggest_categorical('stacking__knn__weights', ['uniform', 'distance']),
        'stacking__knn__metric': lambda trial: trial.suggest_categorical('stacking__knn__metric', ['euclidean', 'manhattan']),
        
        # Meta-learner (LogisticRegression) hyperparameters
        'stacking__final_estimator__C': lambda trial: trial.suggest_float('stacking__final_estimator__C', 1e-4, 1e2, log=True),
        'stacking__final_estimator__solver': lambda trial: trial.suggest_categorical('stacking__final_estimator__solver', ['lbfgs', 'saga']),
        'stacking__final_estimator__max_iter': lambda trial: trial.suggest_categorical('stacking__final_estimator__max_iter', [1000, 2000])
    },
    
    'Bag-SVM': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # Base SVM estimator parameters
        'bagging__estimator__C': lambda trial: trial.suggest_float('bagging__estimator__C', 1e-2, 1e3, log=True),
        'bagging__estimator__kernel': lambda trial: trial.suggest_categorical('bagging__estimator__kernel', ['rbf', 'poly', 'linear']),
        'bagging__estimator__gamma': lambda trial: trial.suggest_categorical('bagging__estimator__gamma', ['scale', 'auto']),
        
        # Bagging ensemble parameters
        'bagging__n_estimators': lambda trial: trial.suggest_int('bagging__n_estimators', 10, 99),
        'bagging__max_samples': lambda trial: trial.suggest_float('bagging__max_samples', 0.5, 0.9),
        'bagging__max_features': lambda trial: trial.suggest_float('bagging__max_features', 0.5, 0.9),
        'bagging__bootstrap': lambda trial: trial.suggest_categorical('bagging__bootstrap', [True, False])
    },
    
    'Bag-KNN': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # Base KNN estimator parameters
        'bagging__estimator__n_neighbors': lambda trial: trial.suggest_int('bagging__estimator__n_neighbors', 3, 20),
        'bagging__estimator__weights': lambda trial: trial.suggest_categorical('bagging__estimator__weights', ['uniform', 'distance']),
        'bagging__estimator__metric': lambda trial: trial.suggest_categorical('bagging__estimator__metric', ['euclidean', 'manhattan', 'minkowski']),
        
        # Bagging ensemble parameters
        'bagging__n_estimators': lambda trial: trial.suggest_int('bagging__n_estimators', 10, 99),
        'bagging__max_samples': lambda trial: trial.suggest_float('bagging__max_samples', 0.5, 0.9),
        'bagging__max_features': lambda trial: trial.suggest_float('bagging__max_features', 0.5, 0.9),
        'bagging__bootstrap': lambda trial: trial.suggest_categorical('bagging__bootstrap', [True, False])
    },
    
    'Ada-SVM': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # Base SVM estimator parameters (linear kernel for AdaBoost compatibility)
        'adaboost__estimator__C': lambda trial: trial.suggest_float('adaboost__estimator__C', 1e-2, 1e2, log=True),
        
        # AdaBoost ensemble parameters
        'adaboost__n_estimators': lambda trial: trial.suggest_int('adaboost__n_estimators', 30, 149),
        'adaboost__learning_rate': lambda trial: trial.suggest_float('adaboost__learning_rate', 0.5, 2.0),
        'adaboost__algorithm': lambda trial: trial.suggest_categorical('adaboost__algorithm', ['SAMME'])
    },
    
    'Ada-KNN': {
        # PCA HYPERPARAMETERS (preprocessing layer optimization)
        'pca__n_components': lambda trial: trial.suggest_float('pca__n_components', 0.90, 0.99),
        'pca__whiten': lambda trial: trial.suggest_categorical('pca__whiten', [True, False]),
        'pca__svd_solver': lambda trial: trial.suggest_categorical('pca__svd_solver', ['auto', 'full']),

        # Base KNN estimator parameters
        'adaboost__estimator__n_neighbors': lambda trial: trial.suggest_int('adaboost__estimator__n_neighbors', 3, 20),
        'adaboost__estimator__weights': lambda trial: trial.suggest_categorical('adaboost__estimator__weights', ['uniform', 'distance']),
        'adaboost__estimator__metric': lambda trial: trial.suggest_categorical('adaboost__estimator__metric', ['euclidean', 'manhattan']),
        
        # AdaBoost ensemble parameters
        'adaboost__n_estimators': lambda trial: trial.suggest_int('adaboost__n_estimators', 30, 149),
        'adaboost__learning_rate': lambda trial: trial.suggest_float('adaboost__learning_rate', 0.5, 2.0),
        'adaboost__algorithm': lambda trial: trial.suggest_categorical('adaboost__algorithm', ['SAMME'])  # SAMME works better with KNN
    }
}

**5. Define the score function and the objective function to be used during training**

Let's define the score function that will be used to measure each model's performance. Instead of using just one metric alone, the function enables us to define a weighted multi-metric approach, defining which metrics would be more important for the model's performance. The chosen score is a combination of three important metrics: 

- Recall measures how good the model is at not having false negatives;
- Precision measures how good the model is at not having false positives;  
- Accuracy measures how good the model is at not having false classifications;

In financial transaction risk assessments, it is more important to have fewer false negatives than false positive classifications, because it would be less risky to block a transaction wrongly considered illicit than to not block a transaction wrongly considered licit.

In [94]:
def create_objective(model_name, pipeline, param_dist, X_train, y_train, cv, scorer):
    """Create Optuna objective function for hyperparameter optimization"""
    def objective(trial):
        # Get parameter suggestions by calling lambdas with trial
        params = {}
        for param_name, suggest_fn in param_dist[model_name].items():
            params[param_name] = suggest_fn(trial)
        
        # Set pipeline parameters
        pipeline.set_params(**params)
        
        # Perform cross-validation
        scores = cross_val_score(pipeline, X_train, y_train, cv=cv, scoring=scorer, n_jobs=-1)
        
        # Return mean F1 score
        return scores.mean()
    
    return objective


def aml_composite_score(y_true, y_pred):
    """Custom scoring balancing multiple AML metrics"""
    #accuracy = accuracy_score(y_true, y_pred)
    #recall = recall_score(y_true, y_pred, pos_label='1')  # Illicit detection
    #precision = precision_score(y_true, y_pred, pos_label='1') # Illicit detection
    f1 = f1_score(y_true, y_pred, pos_label='1')  # Illicit detection   

    # Weighted combination emphasizing illicit detection
    #return 0.6 * recall + 0.2 * precision + 0.2 * accuracy
    return f1

composite_scorer = make_scorer(aml_composite_score)

**6. Execute the training** [CAN BE SKIPPED ~ 2h]

Let's execute the training phase using random grid search and execute it in parallel, with cross-validation of all dataset splits (folds) and rank the best models by score function.

The final plot will display all model training samples with their mean and variance performance during training.

In [ ]:
# Training parameters
n_trials = 2  # PARAMETER: Number of Optuna trials per model
n_jobs = -1    # PARAMETER: Use all available cores

# Filter to train only basic models (first 10)
basic_pipelines = pipelines[:10]

# Create result collections
training_models = []
cv_results_all = []  # Store all CV scores for boxplot
study_results = {}   # Store Optuna studies for analysis

print("🔍 Training Models with Optuna Bayesian Optimization...")
print(f"Training {len(basic_pipelines)} models with {n_trials} trials each")
print("-" * 60)

for name, pipe in basic_pipelines:
    print(f"Training {name}...", end=" ", flush=True)
    
    # Create Optuna study with TPE sampler
    study = optuna.create_study(
        direction='maximize',
        sampler=TPESampler(seed=random_seed)
    )
    
    # Create objective function
    objective = create_objective(
        model_name=name,
        pipeline=pipe,
        param_dist=param_distributions,
        X_train=X_train,
        y_train=y_train,
        cv=cv,
        scorer=composite_scorer
    )
    
    # Optimize
    study.optimize(objective, n_trials=n_trials, show_progress_bar=False)
    
    # Extract best parameters and set them
    best_params = study.best_params
    pipe.set_params(**best_params)
    
    # Get CV scores for evaluation (doesn't fit the original pipe!)
    cv_scores = cross_val_score(pipe, X_train, y_train, cv=cv, scoring=composite_scorer, n_jobs=n_jobs)
    cv_results_all.append(cv_scores)
    
    # Explicitly fit pipeline on full training set to avoid NotFittedError
    pipe.fit(X_train, y_train)
    
    best_score = study.best_value
    std_score = cv_scores.std()
    
    print(f"✅ {best_score:.4f} (±{std_score:.4f})")
    
    training_models.append((name, pipe))
    study_results[name] = study

# Create boxplot with individual CV fold scores
fig = plt.figure(figsize=(25,6))
fig.suptitle('Trained Models Comparison - CV Score Distribution (Optuna Optimization)') 
ax = fig.add_subplot(111) 
plt.boxplot(cv_results_all, labels=[name for name, _ in training_models])
ax.set_ylabel('Cross-Validation F1 Score')
ax.tick_params(axis='x', rotation=45)
plt.show()

optimized_models = training_models.copy()

🔍 Training Models with Optuna Bayesian Optimization...
Training 9 models with 2 trials each
------------------------------------------------------------
Training Ada... ✅ 0.7567 (±0.0090)
Training GB... ✅ 0.8600 (±0.0056)
Training Vote-Soft... ✅ 0.8510 (±0.0024)
Training Vote-Hard... ✅ 0.7985 (±0.0033)
Training Stack... ✅ 0.8172 (±0.0051)
Training Bag-SVM... ✅ 0.8017 (±0.0052)
Training Bag-KNN... ✅ 0.8296 (±0.0030)
Training Ada-SVM... 

**Optuna Optimization Analysis**

Visualize Optuna optimization history and parameter importance for each model.

In [ ]:
# Visualize Optuna optimization for top 3 models
top_models = sorted(training_models, key=lambda x: study_results[x[0]].best_value, reverse=True)[:3]

print("📊 Optuna Optimization Analysis for Top 3 Models")
print("-" * 60)

for name, _ in top_models:
    study = study_results[name]
    print(f"\n{name} - Best F1 Score: {study.best_value:.4f}")
    print(f"Best Parameters: {study.best_params}")
    
    # Plot optimization history
    fig = optuna.visualization.plot_optimization_history(study)
    fig.update_layout(title=f"{name} - Optimization History")
    fig.show()
    
    # Plot parameter importances
    try:
        fig = optuna.visualization.plot_param_importances(study)
        fig.update_layout(title=f"{name} - Parameter Importances")
        fig.show()
    except:
        print(f"  (Not enough trials for parameter importance analysis)")
    
print("\n" + "=" * 60)

**7. Save or load all resulting models**

Let's save all resulting trained models locally or retrieve previously trained models from local or remote sources.

In [ ]:
# Save or Load models
dataset_str  = "mvp-kyt-sup-main"
models_str  = "models"
folder_str = f"./{models_str}/{dataset_str}"
try:
    os.makedirs(folder_str, exist_ok=True)
    for name, pipe in optimized_models:
        joblib.dump(pipe, f"{folder_str}/{name}.pkl", compress=True)
    print(f"💾 Saved {len(optimized_models)} models: {[name for name, _ in optimized_models]}")
except:
    if azureClient.download_documents(models_str, dataset_str, base_path="./"):
        print("Download from Azure Blob Storage completed successfully.")

    folder_dir = Path(folder_str)
    if(folder_dir.exists() and any(folder_dir.iterdir())):
        print("❌ No models were trained.")
        optimized_models = []
        if os.path.exists(folder_dir):
            for file in os.listdir(folder_dir):
                if file.endswith('.pkl'):
                    name = file.replace('.pkl', '')
                    pipe = joblib.load(f"{folder_str}/{file}")
                    optimized_models.append((name, pipe))
            print(f"📁 Loaded {len(optimized_models)} models: {[name for name, _ in optimized_models]}")
        else:
            print("❌ No models found")
    else:
        print("❌ No models available.")

**8. Validate all models and select the best models**

Let's validate and select the best models by applying all trained pipelines to the previously generated testing set using the multi-metric score function.    

Model validation with an unseen dataset during training can give us an approximate measure of how the model would perform in the real world.

In [ ]:
# Select best models based on test accuracy
test_results = []
for name, pipe in optimized_models:
    y_pred = pipe.predict(X_test)
    accuracy = aml_composite_score(y_test, y_pred)
    test_results.append((name, accuracy))

# Sort by test accuracy and get top 3
test_results.sort(key=lambda x: x[1], reverse=True)
print(f"\n🏆 Final top models:")
print('-'*30)
for name, acc in test_results:
    print(f"{name}: Test performance = {acc:.4f}")

**9. Use best models to predict unknown data**

Let's apply the best pipeline model to unknown data—data that does not have labels—and display the results to get an idea of how the landscape of unknown illicit transactions could be. This measure can also be used for comparison with future improvements to the machine learning techniques.

In [ ]:
# Apply best model to unlabeled data
best_model_name = test_results[0][0]
best_model = next(model for name, model in optimized_models if name == best_model_name)
X_unlabeled = df_unlabeled.drop(['class', 'txId'], axis=1)
predictions = best_model.predict(X_unlabeled)
df_prediction = pd.Series(predictions, name="prediction")
df_final = pd.concat([df_unlabeled[['txId']], df_prediction], axis=1)
df_final = df_final.applymap(lambda x: 'Illicit' if x == '1' else 'Licit' if x == '2' else x)

# Analyze prediction distribution
class_counts = df_final['prediction'].value_counts()
labeled_only = class_counts[class_counts.index != 'unknown']
imbalance_ratio = labeled_only.max() / labeled_only.min() if len(labeled_only) >= 2 else 1.0

# Plot distribution
print(f"\n📈 Prediction Distribution:")
print(f"Model used:", best_model_name)
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 4))
class_counts.plot(kind='bar', ax=ax1, color=['lightblue', 'orange', 'lightcoral'])
ax1.set_title('Class Counts')
ax1.tick_params(axis='x', rotation=45)
class_counts.plot(kind='pie', ax=ax2, autopct='%1.1f%%', colors=['lightblue', 'orange', 'lightcoral'])
ax2.set_title('Class Distribution')
ax2.set_ylabel('')
plt.tight_layout()
plt.show()

# Display prediction samples and summary statistics
print(f"📊 Prediction Summary:")
print(f"  Total predictions: {len(df_final):,}")
print(f"  Illicit transactions: {sum(df_final['prediction'] == 'Illicit'):,}")
print(f"  Licit transactions: {sum(df_final['prediction'] == 'Licit'):,}")
print(f"  Imbalance ratio: {imbalance_ratio:.1f}:1")

# Get sample transactions for analysis
illicit_selector = df_final['prediction'] == 'Illicit'
X_unlabeled_illicit = df_final[illicit_selector].head(100)
print(f"\n🔮 Sample illicit transactions (showing {len(X_unlabeled_illicit)} of {sum(illicit_selector):,} total)")
display(X_unlabeled_illicit)

licit_selector = df_final['prediction'] == 'Licit'
X_unlabeled_licit = df_final[licit_selector].head(100)
print(f"\n🔮 Sample licit transactions (showing {len(X_unlabeled_licit)} of {sum(licit_selector):,} total)")
display(X_unlabeled_licit)

### Considerations

Some considerations must be made before the conclusion:

- Compressed data can increase training time but reduces dataset sizes without influencing the final result.

- Models like Naive Bayes reached almost 0.6 in contrast to other models and were removed from training;  

- Models like Logistic Regression also had poor performance and were removed from training, but they are used in some ensembles as default estimator models. 

**Future improvements to training:**

- The unsupervised approach could produce good or even better results because it would use much more data to identify patterns by using the complete dataset. Also, the labeled dataset could indicate which clusters could be labeled with the illicit class;

- The training could use models more recommended for graph-type datasets, such as Graph Convolutional Networks (GCN), making use of the edge dataset to learn patterns with deeper transaction chains, not only direct neighbors.

**Production readiness:**

- This training was performed on a dataset curated for research purposes. There is no information about which features were used, so in order to have a production-ready model, a new dataset in the same format would need to be gathered and curated;

- A final performance indicator would need to be established to consider the model ready for a production environment, by classifying real labeled current data.

### Conclusions

#### Model Performance Summary

This supervised learning approach successfully developed a high-performance KYT system achieving **88.49% performance** on cryptocurrency transaction risk classification. The **SVM (Support Vector Machine)** emerged as the champion model, demonstrating superior performance in distinguishing illicit from licit Bitcoin transactions.

#### Key Technical Achievements

- **Dimensionality Reduction**: PCA preprocessing reduced feature space from 166 to 59 dimensions while preserving 95% variance
- **Algorithm Comparison**: Comprehensive evaluation of 10 ML algorithms with hyperparameter optimization via RandomizedSearchCV  
- **Model Ranking**: SVM (88.49%) > KNN (86.87%) > GB (85.02%) demonstrated that ensemble and kernel methods excel in financial pattern recognition
- **Pipeline Standardization**: StandardScaler + PCA + model architecture ensures consistent preprocessing across algorithms
- **Model Persistence**: All trained models saved with compression for deployment scalability
- **Performance Validation**: Stratified cross-validation ensures reliable performance on imbalanced financial data

#### Real-World Impact

The trained model successfully processed **157,205 unlabeled transactions**, identifying **12,675 potentially illicit transactions**, providing risk assessment capabilities for unknown data—essential for AML compliance.